frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1305,6 +1305,114 @@ const IR::Node* TypeInference::postorder(IR::Concat* expression) {
     return expression;
 }
 
+const IR::Node* TypeInference::postorder(IR::Key* key) {
+    LOG1("TypeInference: visiting " << dbp(key));
+    // compute the type and store it in typeMap
+    auto vec = new IR::Vector<IR::Type>();
+    for (auto ke : *key->keyElements) {
+        auto kt = typeMap->getType(ke->expression);
+        if (kt == nullptr) {
+            LOG2("Bailing out for " << dbp(ke));
+            return key;
+        }
+        vec->push_back(kt);
+    }
+    auto keyTuple = new IR::Type_Tuple(Util::SourceInfo(), vec);
+    LOG2("Setting key type to " << dbp(keyTuple));
+    setType(key, keyTuple);
+    // installing also for the original because we cannot tell which one will survive in the ir
+    // \TODO: figure out the rules of survivorship
```

Comment:
The problem stems from the fact that the visitor always passes you a clone of the node, both in preorder and postorder. But you have no idea whether this copy differs from the original because you are changing it or because someone else changed one of its children.

One can design a visitor were there are never any automatic clones. I proposed to add one such visitor implementation, where all mutations must be made only by the user. This would be easier to debug, but would require users writing more code in some circumstances. Today all transforms use at least O(n) space because of this feature, even if they change no nodes!

Alternatively, we could endow the visitor with an upcall when a node is substituted; this could enable us to incrementally update some data structures such as typemaps.